plenoptic/simulate/models/frontend.py

```diff
@@ -1,96 +1,513 @@
-        self.softplus = nn.Softplus()
+        self.activation = activation
 
-        self.disk_mask = disk_mask
-        self._disk = None  # cached disk to apply to image
+    def forward(self, x: Tensor) -> Tensor:
+        linear = self.center_surround(x)
+        lum = self.luminance(x)
+        lum_normed = linear / (1 + self.luminance_scalar * lum)
+
+        con = self.contrast(lum_normed.pow(2)).sqrt() + 1E-6  # avoid div by zero
+        con_normed = lum_normed / (1 + self.contrast_scalar * con)
+        y = self.activation(con_normed)
+        return y
+
+    def display_filters(self, zoom=5.0, **kwargs):
+        """Displays convolutional filters of model
+
+        Parameters
+        ----------
+        zoom: float
+            Magnification factor for po.imshow()
+        **kwargs:
+            Keyword args for po.imshow
+        Returns
+        -------
+        fig: PyrFigure
+        """
+
+        weights = torch.cat(
+            [
+                self.center_surround.filt,
+                self.luminance.filt,
+                self.contrast.filt,
+            ],
+            dim=0,
+        ).detach()
+
+        title = ["linear", "luminance norm", "contrast norm"]
+
+        fig = imshow(
+            weights, title=title, col_wrap=3, zoom=zoom, vrange="indep0", **kwargs
+        )
+
+        return fig
 
-        if pretrained:
-            self._load_pretrained()
 
-        if not requires_grad:  # turn off gradient
-            [p.requires_grad_(False) for p in self.parameters()]
+class OnOff(nn.Module):
+    """Two-channel on-off and off-on center-surround model with local contrast and
+    luminance gain control.
 
-    def _load_pretrained(self):
-        """Load FrontEnd model weights used from Berardino et al (2017)"""
-        state_dict = torch.load(os.path.join(os.path.dirname(__file__), 'weights/FrontEnd.pt'))
-        self.load_state_dict(state_dict)
+    This model is called OnOff in Berardino et al 2017.
 
-    def _luminance_normalization(self, x):
-        s = self.luminance_scalars
-        return torch.div(self.linear_pad(x), (1 + s * self.luminance_pad(x)))
+    Parameters
+    ----------
+    kernel_size:
+        Shape of convolutional kernel.
+    width_ratio_limit:
+        Ratio of surround stdev over center stdev. Surround stdev will be clamped to
+        ratio_limit times center_std.
+    amplitude_ratio:
+        Ratio of center/surround amplitude. Applied before filter normalization.
+    pad_mode:
+        Padding for convolution, defaults to "circular".
+    activation:
+        Activation function following linear convolution.
+    activation:
+        Activation function following linear and gain control operations.
+    apply_mask:
+        Whether or not to apply circular disk mask to image.
+
+    Notes
+    -----
+    These 12 parameters (standard deviations & scalar constants) were reverse-engineered
+    from model from [1]_, [2]_. Please use at your own discretion.
```

Comment:
Should point out that the "use at your own discretion" only applies if they use the pretrained weights, right?

Reply:
Done